### ISSUE TITLE
Domain Join issue using run_client_join.yml

### ISSUE THREAD

**Issue by Ansible-NOOB** [OneIdentity/ansible-authentication-services#45]
Hello All,

This was working since the last update 3 months ago (and before that for over a year), but it stopped working just recently.

Giving the following error.

argument ''keytab'' is of type <class ''NoneType''> and we were unable to convert to str: ''None'' is not a string and conversion is not allowed

If I specify /etc/opt/quest/vas/hot.keytab, it tells me the key tab is not found.

Hopefully someone will be able to help.

**Comment by erick-OI** [comment]
Can you please let me know if you are using our latest playbooks version 0.3.0? we added the capability to use a keytab just 3 months ago.
https://github.com/OneIdentity/ansible-authentication-services/blob/master/examples/run_client_join.yml

Also, what ansible changes did you perform recently in your environment? I may assist in trying to repro the issue if you provide me with the playbook.

**Comment by Ansible-NOOB** [comment]
Yes I am using the most current version of the playbook, and the most recent version of SAS.

I have verified with my UNIX support team, and they have advised me of no recent changes.

Here is the playbook I am using with my stuff removed. (keytab option is not in it as i have never had to specify.)
---

- hosts: all
  gather_facts: false
  
# The variables you would most likely want/need to override have been included
  vars:

    # Active Directory
    client_join_state: joined
    client_join_extra_args: -c "<Set to a specific OU>"
    client_domain: <mydomain.local>
    client_username: <myADSadminuser>
    client_password: <Thepassword>

    # Facts
    client_join_facts_generate: true
    client_join_facts_verbose: false

    # Reports
    client_join_reports_generate: true
    client_join_reports_backup: false

  roles:
    - name: oneidentity.authentication_services.client_join

**Comment by llnagy76** [comment]
@Ansible-NOOB
Thank you very much for the bug report!

I have tested the client_join role in my older test environment and it works without any problem. In that environment the version of Ansible is 2.9.15 and the version of python is 3.9.9.

Now I have tested the same role in a newer environment, where the version of Ansible is 2.16.2 and the version of python is 3.12.0, and the role has failed with the same error message as in your enviroment.

I have started to work on fixing the bug.

You wrote that 'I have verified with my UNIX support team, and they have advised me of no recent changes.' May I ask you to check if there has been any ansible and / or python upgrade in your environment recently? Thank you!

**Comment by Ansible-NOOB** [comment]
So I confirmed with my UNIX support team, it is confirmed that ansible and python were upgraded as part of their regular patch process on.

Sun 28 Jan 2024 07:40:58 AM MST

Adding the line client_keytab: '' to the run_client_join.yml shoots off a new error. (that even if I run as sudo same issue, which is something I did not have to use before)

TASK [oneidentity.authentication_services.client_join : fail] ******************
fatal: [REDACTED]: FAILED! => {"changed": false, "msg": "Error: You must have root access to run this command."}

TASK [oneidentity.authentication_services.client_join : include_tasks] *********
skipping: [REDACTED]

TASK [oneidentity.authentication_services.client_join : include_tasks] *********
included: /home/myuser/.ansible/collections/ansible_collections/oneidentity/authentication_services/roles/client_join/tasks/utils/set_fact_failed.yml for REDACTED

TASK [oneidentity.authentication_services.client_join : set_fact] **************
ok: [REDACTED]

TASK [oneidentity.authentication_services.client_join : fail] ******************
fatal: [REDACTED: FAILED! => {"changed": false, "msg": "Error: You must have root access to run this command."}

**Comment by llnagy76** [comment]
Join needs root privileges on the managed nodes. See https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_privilege_escalation.html

In my test env I set these variables:

ansible_become: true
ansible_become_password: (pwd)
ansible_user: (username)

I can reproduce your error if I comment out these variables. Do you use these or similar variables to get privilege escalation?

**Comment by Ansible-NOOB** [comment]
Thanks for that, so by adding become: true to my playbook with the workaround supplied, I am now able to once again run the playbook with no errors, and it appears to join the machine to the domain.

I appreciate it!

**Comment by llnagy76** [comment]
@Ansible-NOOB
Thank you for your help in fixing this bug!

**Comment by llnagy76** [comment]
The fix has been released in Release 0.3.1.
